packages/rancher-monitoring/rancher-monitoring.patch

```diff
@@ -706,6 +717,216 @@ diff -x '*.tgz' -x '*.lock' -uNr packages/rancher-monitoring/charts-original/tem
+@@ -1,11 +1,19 @@
+ {{- if and (.Values.alertmanager.enabled) (not .Values.alertmanager.alertmanagerSpec.useExistingSecret) }}
++{{- if .Release.IsInstall }}
++{{- $secretName := (printf "alertmanager-%s-alertmanager" (include "prometheus-operator.fullname" .)) }}
++{{- if (lookup "v1" "Secret" (include "prometheus-operator.namespace" .) $secretName) }}
++{{- required (printf "Cannot overwrite existing secret %s in namespace %s." $secretName (include "prometheus-operator.namespace" .)) "" }}
++{{- end }}{{- end }}
+ apiVersion: v1
+ kind: Secret
+ metadata:
+-  name: alertmanager-{{ template "prometheus-operator.fullname" . }}-alertmanager
++  name: alertmanager-{{ template "prometheus-operator.fullname" . }}-pre-install
+   namespace: {{ template "prometheus-operator.namespace" . }}
+-{{- if .Values.alertmanager.secret.annotations }}
+   annotations:
++    "helm.sh/hook": pre-install
++    "helm.sh/hook-delete-policy": hook-succeeded, hook-failed
++    "helm.sh/hook-weight": "4"
++{{- if .Values.alertmanager.secret.annotations }}
+ {{ toYaml .Values.alertmanager.secret.annotations | indent 4 }}
+ {{- end }}
+   labels:
+@@ -20,4 +28,95 @@
+ {{- range $key, $val := .Values.alertmanager.templateFiles }}
+   {{ $key }}: {{ $val | b64enc | quote }}
+ {{- end }}
++---
++apiVersion: batch/v1
++kind: Job
++metadata:
++  name: alertmanager-{{ template "prometheus-operator.fullname" . }}-pre-install
++  namespace: {{ template "prometheus-operator.namespace" . }}
++  labels:
++{{ include "prometheus-operator.labels" . | indent 4 }}
++    app: {{ template "prometheus-operator.name" . }}-alertmanager
++  annotations:
++    "helm.sh/hook": pre-install
++    "helm.sh/hook-delete-policy": hook-succeeded, hook-failed
++    "helm.sh/hook-weight": "5"
++spec:
++  template:
++    metadata:
++      name: alertmanager-{{ template "prometheus-operator.fullname" . }}-pre-install
++      labels: {{ include "prometheus-operator.labels" . | nindent 8 }}
++        app: {{ template "prometheus-operator.name" . }}-alertmanager
++    spec:
++      serviceAccountName: alertmanager-{{ template "prometheus-operator.fullname" . }}-pre-install
++      containers:
++        - name: copy-pre-install-secret
++          image: {{ template "system_default_registry" . }}{{ .Values.alertmanager.secret.image.repository }}:{{ .Values.alertmanager.secret.image.tag }}
++          imagePullPolicy: {{ .Values.alertmanager.secret.image.pullPolicy }}
++          command:
++          - /bin/sh
++          - -c
++          - >
++            if kubectl get secret -n {{ template "prometheus-operator.namespace" . }} alertmanager-{{ template "prometheus-operator.fullname" . }}-alertmanager > /dev/null 2>&1; then
++              echo "Secret already exists"
++              exit 1
++            fi;
++            kubectl patch secret -n {{ template "prometheus-operator.namespace" . }} --dry-run -o yaml
++            alertmanager-{{ template "prometheus-operator.fullname" . }}-pre-install
```

Comment:
Can you explain this line quick. I'm confused about why we are using `hook-succeeded, hook-failed` on the secret above.

Reply:
Explained here https://github.com/rancher/charts/pull/637#issuecomment-692366050, we want the resources to be deleted regardless of whether the hook is successful or fails since leaving the resources behind on either one of those scenarios will require a manual deletion to retry installing the chart iirc since the resources still exist in the cluster.

Reply:
But that is on the secret, are we immediately deleting the `-pre-install` secret ?

Reply:
Yes, we delete the pre-install secret the moment that it has been copied over into `alertmanager-rancher-monitoring-alertmanager`.

Reply:
Ah, making more sense. So we template the `pre-install` secret, patch/rename on the fly, and kubectl apply that into the `alertmanager` secret, and then delete the former.